### PR TITLE
Helper methods to get charset from Content-Type header of HttpMessage

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Motivation:

Issue #4331:
HttpHeaders already has specific methods for such popular and simple headers like "Host", but if I need to convert POST raw body to string I need to parse complex ContentType header in my code.

Modifications:

Add getCharset and getCharsetAsString methods to parse charset from Content-Length header.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java [hunk 1]
```diff
@@ -313,6 +317,62 @@ public static void setTransferEncodingChunked(HttpMessage m, boolean chunked) {
         }
     }
 
+    /**
+     * Fetch charset from message's Content-Type header.
+     *
+     * @return the charset from message's Content-Type header or {@link java.nio.charset.StandardCharsets#ISO_8859_1}
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: `java.nio.charset.StandardCharsets` only exists in java7. Link to `CharsetUtil` please.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java [hunk 2]
```diff
@@ -313,6 +317,62 @@ public static void setTransferEncodingChunked(HttpMessage m, boolean chunked) {
...
+     * This is not according to standard, but this method provide an ability to catch desired mistakes manually in code
+     *
+     * @return the charset string from message's Content-Type header or {@code null} if charset is not presented
+     */
+    public static String getCharsetAsString(HttpMessage message) {
+        CharSequence contentTypeValue = message.headers().get(HttpHeaderNames.CONTENT_TYPE);
+        if (contentTypeValue == null) {
+            return null;
+        }
+        final String CHARSET_EQUALS = HttpHeaderValues.CHARSET + "=";
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: move this to a private static field.

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.netty.util.internal;
+
+/**
+ * Methods of this class taken from {@link org.apache.commons.lang3.StringUtils} and
+ * {@link org.apache.commons.lang3.CharSequenceUtils} classes of org.apache.commons:commons-lang3 project
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: if this is the case we need to update the NOTICE and LICENSE files.

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+ */
+public final class StringUtils {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: merge all the methods into `StringUtil`. There is no need to create just another class: https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/internal/StringUtil.java

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            return INDEX_NOT_FOUND;
+        }
+        if (startPos < 0) {
+            startPos = 0;
+        }
+        final int endLimit = str.length() - searchStr.length() + 1;
+        if (startPos > endLimit) {
+            return INDEX_NOT_FOUND;
+        }
+        if (searchStr.length() == 0) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: there is no need to call searchStr.length() multiple times. Just call it one time and store it in a local var.

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            return startPos;
+        }
+        for (int i = startPos; i < endLimit; i++) {
+            if (regionMatches(str, true, i, searchStr, 0, searchStr.length())) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: same comment as above

File: common/src/test/java/io/netty/util/AsciiStringCharacterTest.java [hunk 7]
```diff
@@ -233,4 +235,12 @@ public void testBooleanUtilityMethods() {
...
         assertTrue(new AsciiString(new byte[] { 2, 0 }).parseBoolean());
     }
+
+    @Test
+    public void testEqualsIgnoreCase() {
+        assertThat(AsciiString.contentEqualsIgnoreCase(null, null), is(true));
+        assertThat(AsciiString.contentEqualsIgnoreCase(null, "foo"), is(false));
+        assertThat(AsciiString.contentEqualsIgnoreCase("bar", null), is(false));
+        assertThat(AsciiString.contentEqualsIgnoreCase("FoO", "fOo"), is(true));
+    }
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: does this belong to this PR ? Looks not related.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java (hunk 1 above)
Comment (Spikhalskiy) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: ok, will do

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java (hunk 2 above)
Comment (Spikhalskiy) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: ok, will do

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 3 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: Is it a problem? Or I should just make this modification in this PR?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: modify this PR.

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 4 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: In this case, should I put original copyright header on the top of netty's String Util near netty's copyright header or before this methods? Sorry, it's not fully clear for me.

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 5 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: Will fix

File: common/src/test/java/io/netty/util/AsciiStringCharacterTest.java (hunk 7 above)
Comment (Spikhalskiy) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: Just mentioned that this method is in incorrect test class. Can redo, not related.

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+     * StringUtils.indexOfIgnoreCase("abc", "", 9)        = 3
+     * </pre>
+     *
+     * @param str  the CharSequence to check, may be null
+     * @param searchStr  the CharSequence to find, may be null
+     * @param startPos  the start position, negative treated as zero
+     * @return the first index of the search CharSequence (always &ge; startPos),
+     *  -1 if no match or {@code null} string input
+     */
+    public static int indexOfIgnoreCase(final CharSequence str, final CharSequence searchStr, int startPos) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: `AsciiString` already has `containsIgnoreCase`. Can we share code and add `indexOfIgnoreCase` there?

Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: @Scottmitch It was my first wish and I thought about it, bit looks like 1) HttpHeaders values are String. So, CharSequence implementation is String there and not AsciiString, we will need to create AsciiString from them to use this methods. Even if we change this and make them AsciiString - we will need to have something like:
```
if (str instanceof AsciiString) {
   ((AsciiString)str).indexOfIgnoreCase...
} else {
   //call general purpose implementation for CharSequences
}
```
On the same time, AsciiString methods are not general purpose, because they use ByteProcessor and rely on 1 byte - 1 char, so they can't be just moved to String utils and be applied to general CharSequences.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: @Spikhalskiy - `AsciiString` is a `CharSequence` and for lack of better organization that is where a lot of the `CharSequence` utility methods live. We can discuss giving those methods a new home, but for now if may be better to keep them all grouped together.

> AsciiString methods are not general purpose, because they use ByteProcessor and rely on 1 byte - 1 char,

This is an optimization to reduce index offset / checking. `contains` currently does not have any optimizations for `AsciiString`.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java [hunk 9]
```diff
@@ -313,6 +317,62 @@ public static void setTransferEncodingChunked(HttpMessage m, boolean chunked) {
...
+
+    /**
+     * Fetch charset string from message's Content-Type header.
+     *
+     * A lot of sites/possibly clients have charset="CHARSET", for example charset="utf-8". Or "utf8" instead of "utf-8"
+     * This is not according to standard, but this method provide an ability to catch desired mistakes manually in code
+     *
+     * @return the charset string from message's Content-Type header or {@code null} if charset is not presented
+     */
+    public static String getCharsetAsString(HttpMessage message) {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: Is it necessary to force the `toString()` conversion? Can we just return `CharSequence` and if the user needs it in `String` form they can do the `toString()`?

File: common/src/main/java/io/netty/util/internal/StringUtils.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+     *
+     * @param cs the {@code CharSequence} to be processed
+     * @param ignoreCase whether or not to be case insensitive
+     * @param thisStart the index to start on the {@code cs} CharSequence
+     * @param substring the {@code CharSequence} to be looked for
+     * @param start the index to start on the {@code substring} CharSequence
+     * @param length character length of the region
+     * @return whether the region matched
+     */
+    static boolean regionMatches(final CharSequence cs, final boolean ignoreCase, final int thisStart,
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: [AsciiString.regionMatches](https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/AsciiString.java#L830) already exists. Can we just update the AsciiString method, or add a static version of this method?

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 8 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: @Scottmitch I understood what do you mean. Yes, looks like this static methods should definitely be moved from AsciiString to separate Utils class like AsciiOpimizedStringUtils. Ok, for now I will just put this two new general purpose methods into AsciiString.

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 3 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: I'm not convinced we will need to borrow anything.  We have most of the building blocking in `AsciiString` already.

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 10 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: Ok, will try to rework it to static and general purpose version for CharSequence or add new one.

File: common/src/main/java/io/netty/util/internal/StringUtils.java (hunk 3 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: I will take a look once more. I started with reusing methods on AsciiString, but stooped to do it because found them Ascii-ByteArray specific. Will revisit it once more with fresh sight.

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: @Spikhalskiy - Thanks :) Let me know if you have any questions or run into any trouble.

Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/internal/StringUtils.java]: @Scottmitch For example, AsciiString#regionMatches(boolean ignoreCase, int thisStart, CharSequence string, int start, int length) that I need now works only for ASCII in case of ignoreCase=true, because use ascii-only equalsIgnoreCase internally, and it's fine because it's even not static. Will try to rework this methods to static + use CharEqualityComparators to incapsulate charset-related logic to make them general-purpose.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java (hunk 9 above)
Comment (Spikhalskiy) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: sure

**Comment by Spikhalskiy** [comment]
Hi @normanmaurer @Scottmitch, tried to address your comments, please take a look over this PR once more when you get a chance.

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java [hunk 11]
```diff
@@ -315,6 +321,61 @@ public static void setTransferEncodingChunked(HttpMessage m, boolean chunked) {
...
+        }
+        int indexOfCharset = AsciiString.indexOfIgnoreCase(contentTypeValue, CHARSET_EQUALS, 0);
+        if (indexOfCharset != AsciiString.INDEX_NOT_FOUND) {
+            int indexOfEncoding = indexOfCharset + CHARSET_EQUALS.length();
+            if (indexOfEncoding < contentTypeValue.length()) {
+                return contentTypeValue.subSequence(indexOfEncoding, contentTypeValue.length());
+            } else {
+                return null;
+            }
+        } else {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: can you remove this `else` block and just do `return null;`?  Consider removing the previous `else` block too and just fall through to the `return null;` at the end of the method.

File: common/src/main/java/io/netty/util/AsciiString.java [hunk 12]
```diff
@@ -1565,6 +1567,113 @@ private static boolean contains(CharSequence a, CharSequence b, CharEqualityComp
...
+            while (csIndex < csEnd) {
+                char c1 = cs.charAt(csIndex++);
+                char c2 = string.charAt(stringIndex++);
+
+                if (c1 == c2) {
+                    continue;
+                }
+
+                if (ignoreCase) {
+                    //For motivation, why we need two checks, see comment in String#regionMatches
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/AsciiString.java]: Do we need to worry about non-ascii (Georgian alphabet)?

Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/AsciiString.java]: @Scottmitch Interesting question. For new HttpUtil method - definitely no. But in this case this method will become non universal. I think, let's do it this way:
1) Left this methods as it is.
2) Make two more optimized methods - regionMatchesAscii and indexOfAscii.
3) In regionMatchedAscii will be only
```
Character.toLowerCase(c1) != Character.toLowerCase(c2)
```
and I will delete this
```
if (cs instanceof String && string instanceof String) {
    return ((String) cs).regionMatches(ignoreCase, csStart, (String) string, start, length);
}
```
Because currently we will actually go to String implementation and will make two useless checks.

What do you think?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java (hunk 11 above)
Comment (Spikhalskiy) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: sure

File: common/src/main/java/io/netty/util/AsciiString.java (hunk 12 above)
Comment (Spikhalskiy) [inline common/src/main/java/io/netty/util/AsciiString.java]: Reworked to two versions, regionMatches decoupled to decrease code duplication

File: common/src/test/java/io/netty/util/AsciiStringCharacterTest.java [hunk 13]
```diff
@@ -233,4 +235,44 @@ public void testBooleanUtilityMethods() {
...
+        assertEquals(0, AsciiString.indexOfIgnoreCase("", "", 0));
+        assertEquals(0, AsciiString.indexOfIgnoreCase("aabaabaa", "A", 0));
+        assertEquals(2, AsciiString.indexOfIgnoreCase("aabaabaa", "B", 0));
+        assertEquals(1, AsciiString.indexOfIgnoreCase("aabaabaa", "AB", 0));
+        assertEquals(5, AsciiString.indexOfIgnoreCase("aabaabaa", "B", 3));
+        assertEquals(-1, AsciiString.indexOfIgnoreCase("aabaabaa", "B", 9));
+        assertEquals(2, AsciiString.indexOfIgnoreCase("aabaabaa", "B", -1));
+        assertEquals(2, AsciiString.indexOfIgnoreCase("aabaabaa", "", 2));
+        assertEquals(-1, AsciiString.indexOfIgnoreCase("abc", "", 9));
+
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: remove empty line

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java [hunk 14]
```diff
@@ -51,10 +53,38 @@ public void testGetOperations() {
     }
 
     @Test
-    public void testEquansIgnoreCase() {
-        assertThat(AsciiString.contentEqualsIgnoreCase(null, null), is(true));
-        assertThat(AsciiString.contentEqualsIgnoreCase(null, "foo"), is(false));
-        assertThat(AsciiString.contentEqualsIgnoreCase("bar", null), is(false));
-        assertThat(AsciiString.contentEqualsIgnoreCase("FoO", "fOo"), is(true));
+    public void getCharsetAsRawString() {
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java]: Prefix with `test`,

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java [hunk 15]
```diff
@@ -51,10 +53,38 @@ public void testGetOperations() {
...
+        HttpMessage message = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+        message.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/html; charset=\"utf8\"");
+        assertEquals("\"utf8\"", HttpUtil.getCharsetAsString(message));
+
+        message.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/html");
+        assertNull(HttpUtil.getCharsetAsString(message));
+    }
+
+    @Test
+    public void getCharset() {
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java]: Prefix with `test`.

File: codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java [hunk 16]
```diff
@@ -51,10 +53,38 @@ public void testGetOperations() {
...
+        HttpMessage message = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+        message.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/html; charset=utf-8");
+        assertEquals(CharsetUtil.UTF_8, HttpUtil.getCharset(message));
+
+        message.headers().set(HttpHeaderNames.CONTENT_TYPE, "TEXT/HTML; CHARSET=UTF-8");
+        assertEquals(CharsetUtil.UTF_8, HttpUtil.getCharset(message));
+    }
+
+    @Test
+    public void getCharset_defaultValue() {
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/HttpUtilTest.java]: Prefix with `test`.

File: common/src/test/java/io/netty/util/AsciiStringCharacterTest.java [hunk 17]
```diff
@@ -233,4 +235,44 @@ public void testBooleanUtilityMethods() {
...
+    @Test
+    public void testEqualsIgnoreCase() {
+        assertThat(AsciiString.contentEqualsIgnoreCase(null, null), is(true));
+        assertThat(AsciiString.contentEqualsIgnoreCase(null, "foo"), is(false));
+        assertThat(AsciiString.contentEqualsIgnoreCase("bar", null), is(false));
+        assertThat(AsciiString.contentEqualsIgnoreCase("FoO", "fOo"), is(true));
+    }
+
+    @Test
+    public void indexOfIgnoreCase() {
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: Prefix with `test`.

File: common/src/test/java/io/netty/util/AsciiStringCharacterTest.java [hunk 18]
```diff
@@ -233,4 +235,44 @@ public void testBooleanUtilityMethods() {
...
+        assertEquals(-1, AsciiString.indexOfIgnoreCase("aabaabaa", "B", 9));
+        assertEquals(2, AsciiString.indexOfIgnoreCase("aabaabaa", "B", -1));
+        assertEquals(2, AsciiString.indexOfIgnoreCase("aabaabaa", "", 2));
+        assertEquals(-1, AsciiString.indexOfIgnoreCase("abc", "", 9));
+
+        assertEquals(0, AsciiString.indexOfIgnoreCase("ãabaabaa", "Ã", 0));
+    }
+
+    @Test
+    public void indexOfIgnoreCaseAscii() {
```
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: Prefix with `test`.

Comment (Spikhalskiy) [inline common/src/test/java/io/netty/util/AsciiStringCharacterTest.java]: @normanmaurer Done with code style issues

**Comment by normanmaurer** [comment]
@Scottmitch want to have a final look ?

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java [hunk 19]
```diff
@@ -27,6 +31,8 @@
  * Utility methods useful in the HTTP context.
  */
 public final class HttpUtil {
+    private static final String CHARSET_EQUALS = HttpHeaderValues.CHARSET + "=";
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpUtil.java]: Can you make this an AsciiString ?

```java
private static final AsciiString CHARSET_EQUALS = AsciiString.of(HttpHeaderValues.CHARSET + "=");
```

**Comment by normanmaurer** [comment]
@Spikhalskiy just one thing and then we should be ready too go. Please also squash the commits.

**Comment by Spikhalskiy** [comment]
@normanmaurer fixed, rebased, squashed.
@normanmaurer @Scottmitch Thank you for review!

**Comment by CodingFabian** [comment]
are we not tracking which part of a lib we are copying? What if somebody removes these methods, would they know that they have to remove the common slang license reference again?

**Comment by Spikhalskiy** [comment]
Rebased to resolve merge conflict

**Comment by normanmaurer** [comment]
@CodingFabian not really... I just "depend" on my memory here, not sure if this is good or not ;)

**Comment by normanmaurer** [comment]
@Spikhalskiy sorry it took so long...

Cherry-picked into 4.1 (2a65ae256ec2170edaa50c9360a0afbfea15319c) and master (e4a9f055f358a4ea135c1132478ceadf39be01cf)

**Comment by normanmaurer** [comment]
@Spikhalskiy also :heart: ;)
